controllers/object_controls.go

```diff
@@ -773,6 +785,73 @@ func applyCommonDaemonsetConfig(obj *appsv1.DaemonSet, config *gpuv1.ClusterPoli
 	return nil
 }
 
+// apply necessary transforms if a custom host root path is configured
+func transformForHostRoot(obj *appsv1.DaemonSet, hostRoot string) {
+	if hostRoot == "" || hostRoot == "/" {
+		return
+	}
+
+	transformHostRootVolume(obj, hostRoot)
+	transformHostDevCharVolume(obj, hostRoot)
+}
+
+func transformHostRootVolume(obj *appsv1.DaemonSet, hostRoot string) {
+	containsHostRootVolume := false
+	for _, volume := range obj.Spec.Template.Spec.Volumes {
+		if volume.Name == "host-root" {
+			volume.HostPath.Path = hostRoot
+			containsHostRootVolume = true
+			break
+		}
+	}
+
+	if !containsHostRootVolume {
+		return
+	}
+
+	for index := range obj.Spec.Template.Spec.InitContainers {
+		setContainerEnv(&(obj.Spec.Template.Spec.InitContainers[index]), HostRootEnvName, hostRoot)
+	}
+
+	for index := range obj.Spec.Template.Spec.Containers {
+		setContainerEnv(&(obj.Spec.Template.Spec.Containers[index]), HostRootEnvName, hostRoot)
+	}
+}
+
+func transformHostDevCharVolume(obj *appsv1.DaemonSet, hostRoot string) {
+	for _, volume := range obj.Spec.Template.Spec.Volumes {
+		if volume.Name == "host-dev-char" {
+			volume.HostPath.Path = filepath.Join(hostRoot, "/dev/char")
+			break
+		}
+	}
+}
+
+// apply necessary transforms if a custom driver install directory is configured
+func transformForDriverInstallDir(obj *appsv1.DaemonSet, driverInstallDir string) {
+	if driverInstallDir == "" || driverInstallDir == DefaultDriverInstallDir {
+		return
+	}
+
+	podSpec := obj.Spec.Template.Spec
+	for _, volume := range podSpec.Volumes {
+		if volume.Name == "driver-root" {
+			volume.HostPath.Path = driverInstallDir
+		}
+	}
+
+	for i, ctr := range podSpec.InitContainers {
+		if ctr.Name == "driver-validation" {
+			setContainerEnv(&(podSpec.InitContainers[i]), "DRIVER_ROOT", driverInstallDir)
```

Comment:
I may have asked somewhere. What is the difference between `DRIVER_ROOT` and `NVIDIA_DRIVER_ROOT`?

Reply:
For clarity, I have updated the envvars for the driver-validator to `DRIVER_INSTALL_DIR` and `DRIVER_INSTALL_DIR_CTR_PATH` to align with the name of the field in ClusterPolicy. `DRIVER_INSTALL_DIR` represents the path on the host, while `DRIVER_INSTALL_DIR_CTR_PATH` represents the path mounted in the driver-validator container.